src/sidebar/services/annotation-activity.ts

```diff
@@ -43,12 +47,28 @@ export class AnnotationActivityService {
     };
 
     if (this._reportConfig.events.includes(eventType)) {
-      postMessageJsonRpc.notify(
-        this._rpc.targetFrame,
-        this._rpc.origin,
-        this._reportConfig.method,
-        [eventType, data],
-      );
+      this.notify(this._reportConfig.method, [eventType, data]);
     }
   }
+
+  /**
+   * Notify the embedder frame that the client has, or does not have,
+   * unsaved changes to annotations.
+   */
+  notifyUnsavedChanges(unsaved: boolean) {
+    this.notify('reportUnsavedChanges', [{ unsaved }]);
+  }
+
+  /** Send a JSON-RPC message to the embedder frame. */
+  private notify(method: string, args: unknown[]) {
```

Comment:
This would benefit from stronger typing, but I think that's outside the scope of this PR.